src/Microsoft.VisualStudio.ProjectSystem.Managed/Tree/Dependencies/Assets/AssetsFileDependenciesDataSource.cs

```diff
@@ -0,0 +1,112 @@
+                = DataflowBlockSlim.CreateTransformBlock<IProjectVersionedValue<ValueTuple<IProjectSnapshot, IProjectSubscriptionUpdate>>, IProjectVersionedValue<AssetsFileDependenciesSnapshot>>(Transform, skipIntermediateInputData: true, skipIntermediateOutputData: true);
+
+            return new DisposableBag
+            {
+                // Subscribe to "ConfigurationGeneral" rule data
+                projectRuleSource.LinkTo(
+                    intermediateBlock,
+                    ruleNames: ConfigurationGeneral.SchemaName,
+                    suppressVersionOnlyUpdates: false,
+                    linkOptions: DataflowOption.PropagateCompletion),
+
+                // Sync link inputs, joining on versions, and passing joined data to our transform block
+                ProjectDataSources.SyncLinkTo(
+                    _activeConfiguredProjectSnapshotService.SourceBlock.SyncLinkOptions(),
+                    intermediateBlock.SyncLinkOptions(),
+                    transformBlock,
+                    linkOptions: DataflowOption.PropagateCompletion),
+
+                // Flow transformed data to the output/target
+                transformBlock.LinkTo(targetBlock, DataflowOption.PropagateCompletion)
+            };
+
+            IProjectVersionedValue<AssetsFileDependenciesSnapshot> Transform(IProjectVersionedValue<ValueTuple<IProjectSnapshot, IProjectSubscriptionUpdate>> update)
+            {
+                var projectSnapshot = (IProjectSnapshot2)update.Value.Item1;
+                IProjectSubscriptionUpdate subscriptionUpdate = update.Value.Item2;
+
+                string? path = null;
+                DateTime timestampUtc = DateTime.MinValue;
+
+                AssetsFileDependenciesSnapshot snapshot = lastSnapshot;
+
+                if (subscriptionUpdate.CurrentState.TryGetValue(ConfigurationGeneral.SchemaName, out IProjectRuleSnapshot ruleSnapshot))
+                {
+                    if (ruleSnapshot.Properties.TryGetValue(ConfigurationGeneral.ProjectAssetsFileProperty, out path))
+                    {
+                        if (!string.IsNullOrEmpty(path))
+                        {
+                            if (projectSnapshot.AdditionalDependentFileTimes == null || !projectSnapshot.AdditionalDependentFileTimes.TryGetValue(path, out timestampUtc))
+                            {
+                                try
+                                {
+                                    timestampUtc = File.GetLastWriteTimeUtc(path);
+                                }
+                                catch
+                                {
+                                    // ignore
+                                }
+                            }
+
+                            if (path != lastAssetsFilePath || timestampUtc != lastTimestampUtc)
```

Comment:
Be aware that this logic can result in a timestamp that regresses (if file is touched 3 times and evaluation skips the first touch, picks up the second, but you manually read the 3rd), however, the logic seems okay in the face of that.

Reply:
Added some comments around this.